Eliminate some but not all uses of RVD.rdd

danking opened the pull request:
This change anticipates the ContextRDD change wherein `RVD.rdd` will not
be an RDD. Moreover, enforcing an abstraction barrier at the level of
`RVD` will ease changes to the implementation of `RVD`.

There are two remaining types of calls that I cannot eliminate:

 - uses in BlockMatrix and OrderedRDD2: these two classes are building
   new RDDs based on the RVD's rdd, these classes should be considered
   within the implementation of the RVD abstraction. Because these two
   classes are outside of `is.hail.rvd`, I cannot enforce an access
   modifier on `RVD.rdd`.

 - uses by methods:

   - LDPrune: it seems we need a "GeneralRVD"

   - Skat: it seems like some of this could be moved to python actually;
     but there is some matrix math that cannot be moved until the expr
     lang has efficient small-matrix ops

   - MatrixTable.same: I could probably move this if I re-implemented
     forall in terms of RVD.aggregate?

   - MatrixTable.annotateRowsIntervalTable: really not sure about this
     one, this seems like a performance optimization that purposely
     reaches through the abstraction to do Smart Things

cseed commented on src/main/scala/is/hail/rvd/RVD.scala:
```diff
@@ -204,4 +205,26 @@ trait RVD {
+    OrderedRVD.shuffle(typ, partitioner, rdd)
+  }
+
+  def assertOrdered(typ: OrderedRVDType, partitioner: OrderedRVDPartitioner): OrderedRVD = {
```
The other way to write this is a constructor OrderedRVD(RVD, ...).

cseed replied:
I kind of prefer this because it's more scalable in RVD subclasses.

danking replied:
What do you mean by it's more scalable?

cseed replied:
I mean, it seems bad to have methods in the base class for each derived class.

patrick-schultz replied:
In #3159 I added `RVD.constrainToOrderedPartitioner`, which has some overlap with this. It is implemented in `UnpartitionedRVD` by `shuffle`, and in `OrderedRVD` by a narrow-dependency repartition I wrote. It's slightly lower level than this, because it is meant to be used in a case where you want an `OrderedRVD` with *exactly* the given partitioner, dropping any data that falls outside the new partitions.

cseed commented:
1.  Move the classes into is.hail.rvd.  They aren't used anywhere else, are they?

2. @maccum is working over LD prune, the GeneralRDD stuff shouldn't be necessary anymore.  I think she's just working on optimization, so I feel like her current code is an improvement and should go in before the performance improvements are ready.

3. I agree, but, yeah, SKAT can't go until we get expr ndarray.

4. MatrixTable.same just uses zip, you can rewrite it in terms of zip based on my above comments.  You might also need a version of zip that returns a RDD[T].  Compare RVD.map and RVD.mapPartitions.

5. I think the right thing here is an OrderedRVD.orderedIntervalJoin.  Compare @patrick-schultz recent OrderedRVD.orderedJoin (pending): https://github.com/hail-is/hail/pull/3159/files#diff-b173fb9bd584d50afcfa6724954ef3b5R127

cseed commented:
Don't feel you need to do all of these now, just whatever is necessary to minimize the complexity of the ContextRDD and offset-to-pointer changes.

danking commented on src/main/scala/is/hail/expr/Relational.scala:
```diff
@@ -252,42 +252,43 @@ case class MatrixRead(
-                  hn
-                }
+          // FIXME: DK added this assert when refactoring, but should this just
+          // always be true? Is this assert necessary?
```
Yeah the wording of my FIXME is confusing. I was more confused to see `typ.orvdType` at all, it seems like it should always be `rowsRVD.typ`. Should this assert actually be checked immediately after reading? I guess the question is: when is this invariant true? I think maybe it should also be asserted in the constructor of `MatrixValue`.

cseed replied:
Yes, always true.  I would support an assert there, too.  (Although I think MatrixValue shouldn't have a `typ`, but that's a minor thing that I'm not pushing now.)

danking replied:
I agree as well, but also below my needs-change-noise threshold at this point.

danking commented on src/main/scala/is/hail/rvd/OrderedRVD.scala:
```diff
@@ -38,6 +38,9 @@ class OrderedRVD private(
   }
 
+  def unsafeChangeType(newTyp: OrderedRVDType): OrderedRVD =
+    OrderedRVD(newTyp, partitioner, rdd)
```
I guess it's no less unsafe than `mapPreservesPartitioning` in that you could provide a type incompatible with the contents of the resulting OrderedRVD. What do you think of the name `updateType`?

I added this to enable `renameFields` in which, I thought, the whole point was to change the `rowType` (albeit only the names of the fields).

cseed replied:
updateType is a fine name.  I'm not objecting to the operation.  Ah, rename fields.  Hmm.  We don't have something that asserts that the types have the same physical rep.  If TStruct was a complex type with representation TTuple, we could just compare the fundamental types.  I guess that's a mark against my proposal to have TBaseStruct.  Anyway, don't sweat the assert.

danking commented on src/main/scala/is/hail/rvd/OrderedRVD.scala:
```diff
@@ -367,27 +370,46 @@ class OrderedRVD private(
       sc.emptyRDD[RegionValue])
   }
 
+  // FIXME: delete this, it's just a wrapper around coerce
```
I think this is a point of contention, but I found it confusing that both "trust me this is ordered give me an OrderedRVD" and "please do whatever necessary to order this dataset" were both named `apply`.

patrick-schultz replied:
This definitely confused me.

cseed replied:
Agreed, I retract my comment.

danking commented on src/main/scala/is/hail/rvd/OrderedRVD.scala:
```diff
@@ -367,27 +370,46 @@ class OrderedRVD private(
+    OrderedRVD(
+      typ,
+      partitioner,
+      this.rdd.zipPartitions(that, preservesPartitioning)(zipper))
```
It seems like the type should be given by `this.rowTyp ++ that.rowTyp`, not provided, right? If there exist conflicting fields the user can use `updateType` to resolve that.

cseed replied:
Your zipPartitions functions take a f: (Iterator[RV], Iterator[RV]) => Iterator[RV].  The type won't be the concat, but will depend on f.  If you want to concat instead of taking f, then this.rowTyp ++ that.rowTyp seems reasonable.

patrick-schultz replied:
I don't see how not taking a `zipper: (Iterator[RV], Iterator[RV]) => Iterator[RV]` is an option. Then you have to choose how to combine iterators, e.g. some kind of join, and then it's no longer a `zipPartitions` method. But then, as Cotton says, they need to take a type for the region values produced by `zipper`.

I used a similar pattern in #3159 to make `OrderedRVD.join` return an `OrderedRVD`. And when this goes in, I should refactor the join methods to use your `OrderedRVD.zipPartitions`.

danking commented:
Actually I forgot to clean up zipPartitions.

danking commented:
@cseed Ok, I think this should go in if you like the way I addressed your comments. I'll deal with 1-5 as I come across them. There's no current need for RVD.zip so I'll leave that off until I see I need.

danking commented:
@cseed ping, stacked PRs have proven unsuccessful in the past, so I'm keeping the rest of my work gated, but I'd like to start moving things into master. Can you take a look at the latest changes and confirm if you're cool punting on points 1-5 until later?

cseed commented on src/main/scala/is/hail/methods/FilterAlleles.scala:
```diff
@@ -171,17 +171,25 @@ object FilterAlleles {
+          filter(
+            vsm.rvd,
+            removeLeftAligned = true,
+            removeMoving = false,
```
I'm confused how these can be written symmetrically.  One needs a shuffle and the other doesn't.

danking replied:
Definitely a mistake, I didn't preserve the shuffle. Will fix.

What's the difference between coerce and shuffle? Is shuffle just a fast-path (skip checking the partition keys) to the slowest (but most complete) type of coercion?

cseed commented on src/main/scala/is/hail/rvd/OrderedRVD.scala:
```diff
@@ -48,17 +51,17 @@ class OrderedRVD private(
-    OrderedRVD(newTyp,
+  def mapPartitionsPreservesPartitioning(rowTyp: TStruct)
+    (f: (Iterator[RegionValue]) => Iterator[RegionValue])
+      : OrderedRVD =
```
Is this a thing we're doing now, putting the colon on a line by itself?  I thought we liked "Patrick's" style:

```
  def (
    a1: T1,
    ...
    an: Tn
  ): RT
```

I'm not saying no, I'm just asking if we're changing/endorsing a new style

danking replied:
just a formatting flub

cseed commented on src/main/scala/is/hail/rvd/OrderedRVD.scala:
```diff
@@ -367,28 +370,36 @@ class OrderedRVD private(
+  def zipPartitionsPreservesPartitioning[T: ClassTag](
+    newTyp: OrderedRVDType, that: RDD[T]
+  )(zipper: (Iterator[RegionValue], Iterator[T]) => Iterator[RegionValue]
+  ) : OrderedRVD =
```
You have a space before the colon here but not elsewhere.

cseed commented on src/main/scala/is/hail/rvd/OrderedRVD.scala:
```diff
@@ -354,28 +365,36 @@ class OrderedRVD(
-  type CoercionMethod = Int
+  def zipPartitionsPreservesPartitioning[T: ClassTag](
+    newTyp: OrderedRVDType, that: RDD[T]
+  )(zipper: (Iterator[RegionValue], Iterator[T]) => Iterator[RegionValue]
```
Should zipper be on its own line?

cseed replied:
No opinion just asking.

danking replied:
IIRC, the full signature is >80 characters, so I went for the @patrick-schultz style function definition (with each arg on it's own line), but had to improvise a bit since there are two argument lists.

danking replied:
Oh! Yes, I think `that` should also have its own line.

cseed commented on src/main/scala/is/hail/rvd/OrderedRVD.scala:
```diff
@@ -354,28 +365,36 @@ class OrderedRVD(
-  final val SHUFFLE: CoercionMethod = 3
+  def zipPartitionsPreservesPartitioning(newTyp: OrderedRVDType, that: RVD)
+    (zipper: (Iterator[RegionValue], Iterator[RegionValue]) => Iterator[RegionValue])
+      : OrderedRVD =
```
Another dangling colon.

cseed commented on src/main/scala/is/hail/rvd/OrderedRVD.scala:
```diff
@@ -354,28 +365,36 @@ class OrderedRVD(
+      this.rdd.zipPartitions(that.rdd, preservesPartitioning = true)(zipper))
+
+  def writeRowsSplit(path: String, t: MatrixType, codecSpec: CodecSpec)
+      : Array[Long] =
```
And another.

cseed commented on src/main/scala/is/hail/rvd/OrderedRVDType.scala:
```diff
@@ -113,6 +113,12 @@ class OrderedRVDType(
+    partitionKey: Array[String] = partitionKey,
+    key: Array[String] = key,
+    rowType: TStruct = rowType
+  ) : OrderedRVDType = new OrderedRVDType(partitionKey, key, rowType)
```
Space before colon.

cseed commented on src/main/scala/is/hail/rvd/OrderedRVD.scala:
```diff
@@ -23,6 +23,24 @@ class OrderedRVD(
 
+  // should be totally generic, permitting any number of keys, but that requires more work
+  def downcastToPK(): OrderedRVD = {
+    val newType = new OrderedRVDType(partitionKey = typ.partitionKey,
```
You can use OrderedRVDType.copy here.

cseed commented on src/main/scala/is/hail/rvd/OrderedRVD.scala:
```diff
@@ -23,6 +23,24 @@ class OrderedRVD(
+  }
+
+  def upcast(castKeys: Array[String]): OrderedRVD = {
+    val newType = new OrderedRVDType(partitionKey = typ.partitionKey,
```
And here.

cseed commented on src/main/scala/is/hail/variant/MatrixTable.scala:
```diff
@@ -2341,7 +2327,12 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
+
+  def rowsTable(): Table = {
+    val tableType = TableType(rowType, rowKey, globalType)
+    new Table(hc, TableLiteral(TableValue(tableType, globals, rowFieldsRVD)))
```
I'm tempted to ask for a MatrixRowsTable IR node, but I'll hold off for now.

patrick-schultz commented:
Why did you need to add back `downcastToPK` and `upcast`? I tried to make those unnecessary with `KeyedOrderedRVD`.

danking commented:
@patrick-schultz rebase failure, I'll remove.

danking commented:
Removed those, which also addressed cotton's comments. When tests pass I'll merge and then push the next PR up!
